### PR TITLE
Fixed bug with $building.

### PR DESCRIPTION
**Pull request by mj4444ru**
| Q             | A
| ------------- | ---
| Is bugfix?    | ✔️
| New feature?  | ❌
| Breaks BC?    | ❌
| Fixed issues  |

### CODE REVIEW

File: src/Container.php [hunk 1]
```diff
@@ -157,8 +154,11 @@ private function build(string $id)
         }
 
         $this->building[$id] = 1;
-        $object = $this->buildInternal($id);
-        unset($this->building[$id]);
+        try {
+            $object = $this->buildInternal($id);
```
Comment (mj4444ru) [inline src/Container.php:158]: When an error occurred, `$this->building` was not cleared and an error occurred on re-access.

File: src/Container.php [hunk 2]
```diff
@@ -157,8 +154,11 @@ private function build(string $id)
         }
 
         $this->building[$id] = 1;
-        $object = $this->buildInternal($id);
-        unset($this->building[$id]);
```
Comment (mj4444ru) [inline src/Container.php:161]: When an error occurred, `$this->building` was not cleared and an error occurred on re-access.

File: src/Container.php [hunk 3]
```diff
@@ -157,8 +154,11 @@ private function build(string $id)
...
+        try {
+            $object = $this->buildInternal($id);
+        } finally {
+            unset($this->building[$id]);
+        }
```
Comment (mj4444ru) [inline src/Container.php:161]: When an error occurred, `$this->building` was not cleared and an error occurred on re-access.

**Comment by samdark** [comment]
Need a unit test that tests for the behavior changed.

File: tests/Unit/ContainerTest.php [hunk 4]
```diff
@@ -872,6 +872,33 @@ public function register(Container $container): void
         );
     }
 
+    public function testBuildingVar(): void
+    {
+        $building = null;
+        $container = new Container([
+            ColorPink::class => static function (ContainerInterface $container) use (&$building): ColorPink {
+                $building = ((array)$container)["\000Yiisoft\Di\Container\000building"];
```
Comment (samdark) [inline tests/Unit/ContainerTest.php:880]: Is it possible to test it via public API only without accessing internal state?

File: tests/Unit/ContainerTest.php [hunk 5]
```diff
@@ -872,6 +872,33 @@ public function register(Container $container): void
...
+        ]);
+        $container->get(ColorPink::class);
+        $this->assertSame([ColorPink::class => 1], $building);
+    }
+
+    public function testBuildingVarClean(): void
+    {
+        $container = new Container();
+        $container->get(ColorPink::class);
+        $this->assertCount(0, ((array)$container)["\000Yiisoft\Di\Container\000building"]);
```
Comment (samdark) [inline tests/Unit/ContainerTest.php:892]: Is it possible to test it via public API only without accessing internal state?

**Review by samdark — APPROVED** [review]
The fix is logically correct but tests are accessing internal state which isn't good at all. I wonder if you had an issue with the public API or was it spotted during reading code?

**Comment by samdark** [comment]
Thank you! Great catch.

**Pull request merged**